ecosystem/indexer/src/indexer/fetcher.rs

```diff
@@ -3,97 +3,176 @@
+/// Under the hood, it fetches TRANSACTION_FETCH_BATCH_SIZE versions in bulk (when needed), and uses that buffer to feed out
+/// In the event it can't fetch, it will keep retrying every RETRY_TIME_MILLIS ms
+async fn fetch_nexts(client: RestClient, starting_version: u64) -> Vec<Transaction> {
+    let mut retries = 0;
+    while retries < MAX_RETRIES {
+        retries += 1;
+
+        let res = client
+            .get_transactions(Some(starting_version), Some(TRANSACTION_FETCH_BATCH_SIZE))
+            .await;
+
+        match res {
+            Ok(response) => {
+                FETCHED_TRANSACTION.inc();
+                return response.into_inner();
+            }
+            Err(err) => {
+                let err_str = err.to_string();
+                // If it's a 404, then we're all caught up; no need to increment the `UNABLE_TO_FETCH_TRANSACTION` counter
+                if err_str.contains("404") {
+                    aptos_logger::debug!(
                             "Could not fetch {} transactions starting at {}: all caught up. Will check again in {}ms.",
                             TRANSACTION_FETCH_BATCH_SIZE,
-                            self.version,
+                            starting_version,
                             RETRY_TIME_MILLIS,
                         );
-                            tokio::time::sleep(Duration::from_millis(RETRY_TIME_MILLIS)).await;
-                            continue;
-                        }
-                        UNABLE_TO_FETCH_TRANSACTION.inc();
-                        aptos_logger::error!(
-                            "Could not fetch {} transactions starting at {}, will retry in {}ms. Err: {:?}",
-                            TRANSACTION_FETCH_BATCH_SIZE,
-                            self.version,
-                            RETRY_TIME_MILLIS,
-                            err
-                        );
-                        tokio::time::sleep(Duration::from_millis(RETRY_TIME_MILLIS)).await;
-                    }
-                };
+                }
+                UNABLE_TO_FETCH_TRANSACTION.inc();
+                aptos_logger::error!(
+                    "Could not fetch {} transactions starting at {}, will retry in {}ms. Err: {:?}",
+                    TRANSACTION_FETCH_BATCH_SIZE,
+                    starting_version,
+                    RETRY_TIME_MILLIS,
+                    err
+                );
             }
         }
-        // At this point we're guaranteed to have something in the buffer
-        let transaction = transactions_buffer.pop().unwrap();
-        self.version += 1;
-        transaction
+        tokio::time::sleep(Duration::from_millis(RETRY_TIME_MILLIS)).await;
+    }
+
+    panic!(
+        "Could not fetch {} transactions starting at {}!",
+        TRANSACTION_FETCH_BATCH_SIZE, starting_version
+    );
+}
+
+#[derive(Debug)]
+pub struct TransactionFetcher {
+    starting_version: u64,
```

Comment:
Given we don't update the starting_version, why not just pass it as a param in start()?